drivers/mpu_9250.cpp

```diff
@@ -0,0 +1,514 @@
+// PUBLIC FUNCTIONS IMPLEMENTATION
+//------------------------------------------------------------------------------
+Mpu_9250::Mpu_9250(Spi& spi, Gpio& nss_gpio, const conf_t config):
+    spi_(spi),
+    nss_(nss_gpio),
+    acc_data_(std::array<float,3>{{0.0f, 0.0f, 0.0f}}),
+    gyro_data_(std::array<float,3>{{0.0f, 0.0f, 0.0f}}),
+    mag_data_(std::array<float,3>{{0.0f, 0.0f, 0.0f}}),
+    temperature_(0.0f),
+    last_update_us_(0.0f),
+    config_(config)
+{
+    switch (config.acc_range)
+    {
+        case ACC_2G:
+            acc_scale_ = 16384;
+            break;
+
+        case ACC_4G:
+            acc_scale_ = 8192;
+            break;
+
+        case ACC_8G:
+            acc_scale_ = 4096;
+            break;
+
+        case ACC_16G:
+            acc_scale_ = 2048;
+            break;
+    }
+
+    switch (config.gyro_range)
+    {
+        case GYRO_250_DEG:
+            gyro_scale_ = 131;
+            break;
+
+        case GYRO_500_DEG:
+            gyro_scale_ = 65.5;
+            break;
+
+        case GYRO_1000_DEG:
+            gyro_scale_ = 32.8;
+            break;
+
+        case GYRO_2000_DEG:
+            gyro_scale_ = 16.4;
+            break;
+    }
+
+    // converting from dps to rps
+    gyro_scale_ *= 180.0f/3.14159f;
+}
+
+bool Mpu_9250::init(void)
+{
+    bool success = true;
+
+    nss_.init();
+
+    // Reset accelerometer and gyroscope
+    success &= mpu_reset();
+
+    // Read WhoAmI for Acc and Gyr
+    uint8_t whoami_answer   = 0;
+    success &= read_reg(MPU9250_WHOAMI_REG, &whoami_answer);
+    success &= whoami_answer == MPU9250_WHOAMI_ID ? true : false;
+
+    // Automatic best available clock selection
+    uint8_t power_management_cmd = MPU9250_PWRMGMT_PLL_X_CLK;
+    success &= write_reg(MPU9250_PWR_MGMT_REG, &power_management_cmd);
+
+    // Enable communication with Mag
+    uint8_t mag_communication_en = (uint8_t)(MPU9250_USERCTL_I2C_MST_EN |
+                                             MPU9250_USERCTL_DIS_I2C);
+    success &= write_reg(MPU9250_USER_CTRL_REG, &mag_communication_en);
+
+    // Reset magnetometer
+    success &= mag_reset();
+
+    // Read WhoAmI for Mag
+    success &= mag_read_reg(AK8963_WHOAMI_REG, &whoami_answer);
+    success &= whoami_answer == AK8963_WHOAMI_ID ? true : false;
+
+    if (!success)
+    {
+        // Returning if device not responding
+        return success;
+    }
+
+    // Configuring Magnetometer
+    success = set_mag_mode();
+
+    // Use slv0 to access mag raw datas (x,y,z and status 2 register)
+    // burst read used
+    uint8_t first_reg       = AK8963_HXL;
+    uint8_t address         = (uint8_t)(MPU9250_AK8963_ADDR | MPU9250_READ_FLAG);
+    uint8_t slv0_ctrl_reg   = (uint8_t)(MPU9250_I2CSLV_EN | 7);
+
+    success &= write_reg(MPU9250_SLV0_REG_REG , &first_reg);
+    success &= write_reg(MPU9250_SLV0_ADDR_REG, &address);
+    success &= write_reg(MPU9250_SLV0_CTRL_REG, &slv0_ctrl_reg);
+
+    // Configuring Accelerometer and Gyroscope
```

Comment:
you could add a note saying that this needs to be done after the configuration of magnetometer to work properly